### PR TITLE
assume char could be unsigned

### PR DESCRIPTION
**Pull request by alandefreitas**
Fixes errors that happen with `char` is unsigned.

MSVC: `/J`
GCC: `-funsigned-char`

### CODE REVIEW

File: test/unit/pct_encoding.cpp [hunk 1]
```diff
@@ -362,7 +362,7 @@ class pct_encoding_test
         auto const check =
             []( string_view s,
                 string_view m0,
-                bool space_to_plus = false)
```
Comment (vinniefalco) [inline test/unit/pct_encoding.cpp:365]: why change this?

File: include/boost/url/authority_view.hpp [hunk 2]
```diff
@@ -113,7 +113,6 @@ class BOOST_SYMBOL_VISIBLE
             char const* cs) noexcept;
 
     // return offset of id
-    BOOST_URL_CONSTEXPR
```
Comment (vinniefalco) [inline include/boost/url/authority_view.hpp:116]: This needs to go in a separate commit

File: include/boost/url/impl/ipv6_address.ipp [hunk 3]
```diff
@@ -59,9 +59,8 @@ struct h16
         std::uint16_t v;
         for(;;)
         {
-            auto d =
-                grammar::hexdig_value(*it);
-            if(d == -1)
+            char d{' '};
```
Comment (vinniefalco) [inline include/boost/url/impl/ipv6_address.ipp:62]: why the space?

Comment (alandefreitas) [reply include/boost/url/impl/ipv6_address.ipp]: This is related to the asan error we had before. It was complaining the variable was not being initialized. 

I've fixed some other things since then and made `hexdig_value` deterministic so this might not be an error anymore. 

I just pushed another commit to check if any compiler still complains about it.

File: include/boost/url/grammar/charset.hpp [hunk 4]
```diff
@@ -470,17 +470,27 @@ constexpr hexdig_chars_t hexdig_chars{};
...
+    {
+        res = 10 + c - 'A';
+        return true;
+    }
+    if (c >= 'a' && c <= 'f')
+    {
+        res = 10 + c - 'a';
+        return true;
+    }
+    res = static_cast<char>(0);
```
Comment (vinniefalco) [inline include/boost/url/grammar/charset.hpp:492]: why are we casting 0? doesn't `res = 0` work?

Comment (alandefreitas) [reply include/boost/url/grammar/charset.hpp:492]: Removed. I don't even remember why I did that. :confused:
